packages/astro/src/vite-plugin-head/index.ts

```diff
@@ -45,9 +45,34 @@ export default function configHeadVitePlugin(): vite.Plugin {
 
 	return {
 		name: 'astro:head-metadata',
+		enforce: 'pre',
+		apply: 'serve',
 		configureServer(_server) {
 			server = _server;
 		},
+		resolveId(source, importer) {
+			if(importer) {
+				// Do propagation any time a new module is imported. This is because
+				// A module with propagation might be loaded before one of its parent pages
+				// is loaded, in which case that parent page won't have the in-tree and containsHead
+				// values. Walking up the tree in resolveId ensures that they do
+				return this.resolve(source, importer, { skipSelf: true }).then(result => {
+					if(result) {
+						let info = this.getModuleInfo(result.id);
+						if(info && getAstroMetadata(info)) {
+							const astro = getAstroMetadata(info)!;
```

Comment:
Can't we save the result of `getAstroMetadata(info)` in a variable and avoid query the same info twice?

Reply:
We could, it's just a property accessor though (`info.meta.astro`), there's no querying, so no performance issue here. I'd do it if you think it improves readability.

Reply:
I think it does improve it slightly so I went ahead and did so